Segregated support of unique index and unique constraint

guilhermeblanco opened the pull request:
Added missing commits: 

* https://github.com/doctrine/dbal/commit/400d73c8a1e4f53459a877c8652811ee22c0913b
* https://github.com/doctrine/dbal/commit/e91811cb3fe3101d7d8ac95c45f8bc1ba9bd913a
* https://github.com/doctrine/dbal/commit/a4ecf053186b11bc31fbff79ef36248500383908
* https://github.com/doctrine/dbal/commit/d7d1c397a35735383b34632d3a67bbc3a3371a45
* https://github.com/doctrine/dbal/commit/c71b3ae0425e5c039b047825dbf8351a773e6785
* https://github.com/doctrine/dbal/commit/2f9da39ec645b54d08c95649c99d7b87a32bdc72
* https://github.com/doctrine/dbal/commit/9aa4e79aab0477b4212b01d65de53193d9559e39

## BC BREAK: Doctrine\DBAL\Schema\Table constructor new parameter

Deprecated parameter `$idGeneratorType` removed and added a new parameter `$uniqueConstraints`.
Constructor changed from:

`__construct($tableName, array $columns = [], array $indexes = [], array $fkConstraints = [], $idGeneratorType = 0, array $options = [])`

To the new constructor:

`__construct($tableName, array $columns = [], array $indexes = [], array $uniqueConstraints = [], array $fkConstraints = [], array $options = [])`

guilhermeblanco committed:
Segregated support of unique index and unique constraint. Added missing commits: 400d73c8a1e4f53459a877c8652811ee22c0913b, e91811cb3fe3101d7d8ac95c45f8bc1ba9bd913a, a4ecf053186b11bc31fbff79ef36248500383908, d7d1c397a35735383b34632d3a67bbc3a3371a45, c71b3ae0425e5c039b047825dbf8351a773e6785, 2f9da39ec645b54d08c95649c99d7b87a32bdc72 and 9aa4e79aab0477b4212b01d65de53193d9559e39

lcobucci commented:
@morozov should be the one checking this. However, as it was mentioned on Slack, I'm pretty sure you should be looking at using DBAL 4.x instead of 3.x

SenseException commented on src/Schema/Table.php:
```diff
@@ -835,34 +791,245 @@ public function __clone()
-        return $this->_options['comment'] ?? null;
+        $constraint->setLocalTable($this);
+
+        if (strlen($constraint->getName()) > 0) {
```
```suggestion
        if ($constraint->getName() !== '') {
```

guilhermeblanco replied:
This method was being moved to follow the logical class declaration order:
* public constants
* protected constants
* private constants
* public static attributes
* protected static attributes
* private static attributes
* public attributes
* protected attributes
* private attributes
* constructor
* static public methods
* static protected methods
* static private methods
* magic methods
* public methods
* protected methods
* private methods
No work was done here.

SenseException replied:
This was due to consistency to a previous of my comments. Because my suggested change wasn't needed and therefore not applied, this one also can stay as it is.

SenseException commented on src/Schema/Table.php:
```diff
@@ -835,34 +791,245 @@ public function __clone()
+     *
+     * @return string The normalized identifier.
+     */
+    private function normalizeIdentifier(?string $identifier)
```
```suggestion
    private function normalizeIdentifier(?string $identifier) : string
```

SenseException commented on src/Schema/Table.php:
```diff
@@ -835,34 +791,245 @@ public function __clone()
+     *
+     * @return Column[]
+     */
+    private function filterColumns(array $columnNames, bool $reverse = false)
```
```suggestion
    private function filterColumns(array $columnNames, bool $reverse = false) : array
```

SenseException commented on src/Schema/Index.php:
```diff
@@ -58,31 +57,27 @@ public function __construct($indexName, array $columns, $isUnique = false, $isPr
      * @throws InvalidArgumentException
      */
-    protected function _addColumn($column)
+    protected function _addColumn(string $column)
```
```suggestion
    protected function _addColumn(string $column) : void
```

SenseException reviewed (changes requested):
* Can we remove underscore prefix in method and property names at least in new classes?
* Since this is a major release branch that had no release yet, can return values be added to methods (at least those with scopes less than public)?
* As mentioned in a comment it looks like there are missing tests.

guilhermeblanco commented on src/Platforms/AbstractPlatform.php:
```diff
@@ -3078,10 +3097,30 @@ public function usesSequenceEmulatedIdentityColumns()
+     *
+     * @return string
+     */
+    public function getSequencePrefix($tableName, $schemaName = null)
```
This is being used in ORM, and was conceptually supposed to be part of sequence name resolution/definition.

Usage:
* https://github.com/doctrine/orm/blob/master/lib/Doctrine/ORM/Mapping/Builder/ValueGeneratorMetadataBuilder.php#L139
* https://github.com/doctrine/orm/blob/master/tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2825Test.php#L62

guilhermeblanco replied:
Added tests

guilhermeblanco commented on src/Schema/Table.php:
```diff
@@ -835,34 +791,245 @@ public function __clone()
-        // For keeping backward compatibility with MySQL in previous releases, table comments are stored as options.
-        $this->addOption('comment', $comment);
+        $mergedNames = array_merge([$this->getName()], $constraint->getColumns());
+        $name        = strlen($constraint->getName()) > 0
```
Constraint name can be null at this stage, so `!== ''` wouldn't work here.

guilhermeblanco commented:
> * Can we remove underscore prefix in method and property names at least in new classes?

There is only one class added to project (`UniqueConstraint`). I've fixed the only method that was using underscore.

> * Since this is a major release branch that had no release yet, can return values be added to methods (at least those with scopes less than public)?

I've moved a lot of methods following logical declarations. I only added 2 public methods and added type hint and return types there.

> * As mentioned in a comment it looks like there are missing tests.

Fixed that adding tests to `AbstractPlatform::getSequencePrefix()`

greg0ire commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
+.idea
```
I'd rather we use [a global gitinore file](https://greg0ire.fr/git-gud/#/23)

guilhermeblanco replied:
Fixed

SenseException reviewed (dismissed):
Looks good. :+1:  Only the coding style issues seem to prevent a full build.

morozov commented on src/Platforms/AbstractPlatform.php:
```diff
@@ -1550,15 +1551,30 @@ public function getCreateTableSQL(Table $table, $createFlags = self::CREATE_INDE
+            }
+
+            foreach ($table->getUniqueConstraints() as $uniqueConstraint) {
+                /** @var UniqueConstraint $uniqueConstraint */
```
Given the type of the collection being iterated over, is this type hint needed?

https://github.com/doctrine/dbal/blob/de1f02b628082bf3956d1d64941b4480efc81be1/src/Schema/Table.php#L710-L712

morozov commented on src/Platforms/AbstractPlatform.php:
```diff
@@ -2365,25 +2377,33 @@ public function getCheckDeclarationSQL(array $definition)
+        }
+
+        $constraintName  = $name->getQuotedName($this);
+        $constraintName  = ! empty($constraintName) ? $constraintName . ' ' : '';
```
The trailing whitespace is not part of the constraint name. Would it make sense to rename `$flags` to something like `$tokens`, push the name there and then join everything?

morozov replied:
Please avoid `empty()`. It's forbidden by PHPStan in master.

morozov commented on src/Platforms/AbstractPlatform.php:
```diff
@@ -3078,10 +3097,30 @@ public function usesSequenceEmulatedIdentityColumns()
+     *
+     * @return string
+     */
+    public function getSequencePrefix($tableName, $schemaName = null)
```
Please use type declarations for new APIs.

morozov commented on src/Platforms/OraclePlatform.php:
```diff
@@ -380,11 +380,11 @@ public function getListSequencesSQL($database)
      * {@inheritDoc}
      */
-    protected function _getCreateTableSQL($table, array $columns, array $options = [])
+    protected function _getCreateTableSQL($tableName, array $columns, array $options = [])
```
Would it make sense to use `$name` for the table name argument? `$columns` and `$options` also belong to the table but don't have this prefix.

guilhermeblanco replied:
I've improved the code into something better. Marking as resolved.

morozov replied:
What do you mean?

guilhermeblanco replied:
Sorry, comment was misplaced. I was referring to constraintName padding a space at the end. This one is resolved actually,.

morozov commented on src/Platforms/SqlitePlatform.php:
```diff
@@ -319,9 +319,9 @@ public function getForeignKeyDeclarationSQL(ForeignKeyConstraint $foreignKey)
      * {@inheritDoc}
      */
-    protected function _getCreateTableSQL($name, array $columns, array $options = [])
+    protected function _getCreateTableSQL($tableName, array $columns, array $options = [])
```
Please revert.

morozov commented on src/Schema/SchemaException.php:
```diff
@@ -145,6 +146,20 @@ public static function sequenceDoesNotExist($sequenceName)
+    public static function uniqueConstraintDoesNotExist($constraintName, $table)
+    {
+        return new self(
+            sprintf("There exists no unique constraint with the name '%s' on table '%s'.", $constraintName, $table),
```
Please use double quotes to quote names (see https://github.com/doctrine/dbal/pull/3536).

guilhermeblanco replied:
All other methods here are using single quotes, I'm just following the standard.

morozov replied:
We need to make sure that the standard set in master is followed as well during merging up.

guilhermeblanco replied:
Done

morozov commented on src/Schema/Table.php:
```diff
@@ -21,14 +24,14 @@ class Table extends AbstractAsset
+    protected $_primaryKeyName = null;
+
+    /** @var UniqueConstraint[] */
+    protected $_uniqueConstraints = [];
```
This should not have the leading underscore.

morozov commented on src/Schema/Table.php:
```diff
@@ -41,18 +44,27 @@ class Table extends AbstractAsset
-    public function __construct($tableName, array $columns = [], array $indexes = [], array $fkConstraints = [], $idGeneratorType = 0, array $options = [])
-    {
+    public function __construct(
+        $tableName,
```
Should this be `string $name`?

guilhermeblanco replied:
This might break somewhere else. No worth doing it here as part of this PR.

morozov replied:
Let's agree that the starter of the discussion resolves it. The very first instruction of the method checks the length of the table name. What else than a string can it be to work? Do you expect consumers to pass an object that implements `__toString()`?

guilhermeblanco replied:
I've ran a test and it works ok on both DBAL and ORM.
Personally, my changes have no correlation to the parameter $name in Table class, and this discussion was outside of scope of this PR (which leads to longer cycle time). But anyway, I've done the change as requested.

morozov commented on src/Schema/Table.php:
```diff
@@ -139,19 +142,39 @@ public function addIndex(array $columnNames, $indexName = null, array $flags = [
+     *
+     * @return self
+     */
+    public function addUniqueConstraint(array $columnNames, ?string $indexName = null, array $flags = [], array $options = [])
```
This could use a return type hint.

morozov commented on src/Schema/Table.php:
```diff
@@ -597,49 +503,74 @@ public function removeForeignKey($constraintName)
+     * @return bool
      */
-    public function getColumns()
+    public function hasUniqueConstraint($constraintName)
```
`(string $name) : bool`

morozov reviewed (commented):
Since this is a new API and a breaking change, I'd expect thew new API to be documented in the PR description and the BC break documented in upgrade notes.

When we decided to release DBAL 3.0 earlier for compatibility with PHP 8, we agreed that it should only have trivial to handle breaking changes. There doesn't seem to be much BC break here, so it should be good IMO.

@guilhermeblanco please make sure all new methods have native parameter and return type declarations.

guilhermeblanco commented:
@morozov all updated. Please merge if appropriate.

morozov commented on src/Schema/Table.php:
```diff
@@ -682,17 +602,15 @@ public function getColumn($columnName)
-
-        return null;
+        return $this->_primaryKeyName !== null
+            ? $this->getIndex($this->_primaryKeyName)
```
Please keep it the old way for more accurate code coverage.

guilhermeblanco replied:
Code coverage works properly if you separate ternary lines (which I did). Also, ternaries are faster than if-guard clauses.

morozov commented on src/Schema/Table.php:
```diff
@@ -704,7 +622,23 @@ public function getPrimaryKeyColumns()
+     *
+     * @return Column[]
+     */
+    public function getForeignKeyColumns()
```
Should it declare the return type?

guilhermeblanco replied:
This was a moved method, it'll break BC by doing it if anyone extends Table.

morozov commented on src/Schema/Table.php:
```diff
@@ -758,6 +692,16 @@ public function getIndexes()
+     *
+     * @return UniqueConstraint[]
+     */
+    public function getUniqueConstraints()
```
Missing return type declaration.

guilhermeblanco replied:
Done

morozov commented on src/Schema/Table.php:
```diff
@@ -835,34 +779,243 @@ public function __clone()
-     * Normalizes a given identifier.
+     * @return int
+     */
+    protected function _getMaxIdentifierLength()
```
Why does the method declaration need to be moved?

guilhermeblanco replied:
This method was being moved to follow the logical class declaration order:

* public constants
* protected constants
* private constants
* public static attributes
* protected static attributes
* private static attributes
* public attributes
* protected attributes
* private attributes
* constructor
* static public methods
* static protected methods
* static private methods
* public methods
* protected methods
* private methods
* magic methods

guilhermeblanco replied:
For a better wording, the rule to thumb is: Non-public methods should always be at the end of the class declaration.

morozov replied:
Let's do it in a separate PR. Ideally, it should be done consistently and enforced by the standard.

In this PR, it unnecessarily complicates the review since it introduces new methods and shuffles around the existing ones.

morozov replied:
@carusogabriel is there something like above in the standard or it could be implemented there?

guilhermeblanco replied:
@morozov is that a requirement to have this merged? I'm trying to make ORM `master` point to `3.0.x` instead of `missed-commits`. I did update some methods (don't think this one in particular) because phpstan was not happy.

carusogabriel replied:
@morozov Yeap, there's: https://github.com/slevomat/coding-standard#slevomatcodingstandardclassesclassstructure-

I don't think that'll make to the next major version of `doctrine/coding-standard`, as that release is already big enough. I'll propose to v9.

morozov replied:
> […] is that a requirement to have this merged?

I believe it is. Otherwise, I cannot wrap my head around the scope of the PR.

guilhermeblanco replied:
I've moved back all methods to their original (not logical) location.

guilhermeblanco committed:
Type hinted Table constructor

carusogabriel commented on tests/Functional/ExceptionTest.php:
```diff
@@ -300,7 +301,7 @@ public function testConnectionExceptionSqLite() : void
 
         // mode 0 is considered read-only on Windows
-        $mode = PHP_OS === 'Linux' ? 0444 : 0000;
+        $mode = in_array(PHP_OS, ['Linux', 'Darwin']) ? 0444 : 0000;
```
What about inverting the logic to simplify?
```suggestion
        $mode = PHP_OS_FAMILY === 'Windows' ? 0444 : 0000;
```

And use a new constant introduced in PHP 7.2: https://php.net/manual/reserved.constants#PHP_OS_FAMILY

carusogabriel commented on tests/Platforms/SqlitePlatformTest.php:
```diff
@@ -824,4 +824,9 @@ public function testGetCreateTableSQLWithColumnCollation() : void
+
+    public function testGetSequencePrefixWithSchema() : void
+    {
+        self::assertEquals('bar__foo', $this->platform->getSequencePrefix('foo', 'bar'));
```
Here should be 
```suggestion
        self::assertSame('bar__foo', $this->platform->getSequencePrefix('foo', 'bar'));
```

Maybe we can propose this stricter check with `phpstan/phpstan-unit`.

carusogabriel commented on tests/Functional/Schema/MySqlSchemaManagerTest.php:
```diff
@@ -46,8 +46,8 @@ public function testSwitchPrimaryKeyColumns() : void
         self::assertCount(2, $primaryKey);
-        self::assertContains('bar_id', $primaryKey);
-        self::assertContains('foo_id', $primaryKey);
+        self::assertArrayHasKey('bar_id', $primaryKey);
+        self::assertArrayHasKey('foo_id', $primaryKey);
```
Why this change?

guilhermeblanco replied:
Because before the methods were returning an array of column names. Now it's returning a HashMap/array indexed by column name => Column

morozov commented on src/Platforms/AbstractPlatform.php:
```diff
@@ -3078,10 +3092,25 @@ public function usesSequenceEmulatedIdentityColumns()
+    /**
+     * Gets the sequence name prefix based on table information.
+     */
+    public function getSequencePrefix(string $tableName, ?string $schemaName = null) : string
```
This looks like a leaky abstraction to me. Especially because it's not used by DBAL itself. Given that all the internally used APIs are public, should ORM implement this logic instead?

guilhermeblanco replied:
ORM is the one that would be leaking abstraction if calling this method.
Conceptually, ORM is telling DBAL "build me a primary key sequence for this table under this schema". The fact that ORM needs to do more than that is a bad abstraction on its own.
DBAL should defer to platforms that once a PK sequence is requested, a sequence named "bar__foo_seq" should be created if schemas are not supported or "bar.foo_seq" if they are.

I migrated the method to the project that owns it. A later stage would be encapsulate it and handle it accordingly internally without exposing this API.

morozov replied:
> ORM is the one that would be leaking abstraction if calling this method.

As far as I understand, what you're saying is ORM already relies on implementation details of DBAL. This is a given fact and it cannot be used to make DBAL _leak_ even more of its implementation details.

The very fact that there's a string in ORM that "bar__foo_seq" is built with "__" defined DBAL and "_" defined in ORM looks horrible.

I would agree that the DBAL abstraction used by ORM needs to be _improved_ but the fact that it's already bad cannot be used as a reason to make it even worse.

morozov reviewed (changes requested):
The addition of `Platform::getSequencePrefix()` is really questionable. Although this code is ported from the former `develop`, I don't see any PR/discussion through which it would get merged there.

guilhermeblanco commented:
@morozov answered you in the specific comment about why it was moved.

morozov commented:
@guilhermeblanco given the lack of progress on this issue in the past months and the fact that we're approaching the release, I'm inclined to move it out of the release milestone.

Please let me know if you plan to finish the work before the release.

morozov commented:
I removed the `3.0.0` and unmarked the PR as a blocker since the primary focus of the `3.0.0` release is PHP 8 support.

guilhermeblanco commented:
There is no further progress on this PR aside for waiting for merge.
We spoke in depth about this issue, and there are more functionality added to this PR than a method moving. The missing part was merging it. I relocated the method `getSequencePrefix` as described as a TODO on ORM (reference: https://github.com/doctrine/orm/blob/2.8.x/lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php#L3380) , and this seems to be cause of this not being merged.
It's no responsibility to ORM to determine the sequence prefix to be used depending of the Platform being used, so ORM is not the place to have this code. If there is further missing parts to the functionality being moved, then at least the method at the proper repository. The fact that this is not landing on DBAL 3.0.0 means that ORM 3.0.0 will not support this major version at all, which is a really bad setback. 

We've spoken already about the role of DBAL and ORM. We have originally established that DBAL was a support library for ORM, not a fully independent library. If this have changed since the agreement in 2009, then it needs to be throughly discussed and determined how that will happen. As it current stands, DBAL is a support library for ORM, so moving the questioned method is valid and no further work is needed.

morozov commented:
I don't see `getSequencePrefix()` being part of a public DB abstraction API. Until this issue is resolved, this won't be merged.

vladyslavstartsev commented:
@morozov it's there

https://github.com/doctrine/dbal/blob/5c7a0e23a5a43dd4a22c79974257d5b815bb0bfc/src/Platforms/AbstractPlatform.php#L3110

`AbstractPlatform::getSequencePrefix()`

morozov commented:
> I don't see `getSequencePrefix()` being part of a public DB abstraction API.

I mean that I don't agree with this method being introduced.

vladyslavstartsev commented:
@guilhermeblanco  can you please rebase? So that version 3.0 can be released.

guilhermeblanco commented:
I'm on it... should be ready later today or tomorrow

morozov commented:
@guilhermeblanco do you need help with this?

guilhermeblanco commented:
I worked on that yesterday, but the amount of caused conflicts (and their pointed blocks) are too hard to resolve.
I feel the easier approach seems to be cherry pick the commits on top of current 3.0.x branch.
If you have time, feel free to start. I can wrap it up tonight if you can't make it to the end.

morozov commented:
@guilhermeblanco I rebased this and reorganized the commits in the way that each individual commit passes `phpcs` thereby eliminating commits like “Fixed code style”.

Please make the needed code changes and do some more cleanup. E.g.: 
1. The “Normalized parameter name across platforms” is no longer relevant since the same was done in #4172 and #4224.
2. Commits like “Type hinted Table constructor” and “Added missing return method” are not really relevant from the code change history and can be combined with others.
3. If “Moved back methods” revert the changes made in the same PR, it's better to not introduce these changes in the first place in order to not cause potential merge conflicts.
4. Normally, each commit should pass all build stages in order to be really atomic. This can be accomplished via `git rebase -i upstream/master exec $cmd` where `$cmd` is `phpcs`, `phpunit`, `phpstan a`, `psalm`, etc. This will help to bisect potential issues and not deal with knowingly broken states. If you don't want to do this, please squash everything.

morozov commented:
@guilhermeblanco given your lack of availability, how do you think we proceed here? Should I just remove `getSchemaPrefix()` and count that you handle it later in the ORM or you wanted to make more changes? Note that we're already 3 weeks behind the schedule and it makes us maintain an extra `2.10.x` release.

SenseException commented:
I'm fine with the changes. If getting `getSequencePrefix()` moved back to ORM is the last step of this PR that needs to be done, then I'm :+1: with this PR and it can be merged after that.

Handling `getSequencePrefix()` in DBAL can be done in a separate PR in coordination with an ORM PR.
